File: dotnet/P0002_AddTwoNumbers.cs
namespace Problems;

public class P0002_AddTwoNumbers
{
    public class ListNode
    {
        public int val;
        public ListNode? next;
        public ListNode(int val = 0, ListNode? next = null)
        {
            this.val = val;
            this.next = next;
        }
    }

    private class Solution
    {
        public ListNode AddTwoNumbers(ListNode? num1, ListNode? num2)
        {
            static ListNode? AddNumbers(ListNode? num1, ListNode? num2, int carryOver) {
                if (num1 is null && num2 is null) {
                    return carryOver > 0 ? new ListNode(carryOver) : null;
                }
                var sum = (num1?.val ?? 0) + (num2?.val ?? 0) + carryOver;
                return new ListNode(sum % 10, AddNumbers(num1?.next, num2?.next, sum > 9 ? 1 : 0));
            }
            
            return AddNumbers(num1, num2, 0) ?? new ListNode();
        }
    }

    private static IEnumerable<object[]> GetTestCases()
    {
        return new [] {
            new object[] {1, 2, 3},
            new object[] {1, 20 , 21},
            new object[] {1, 999, 1000}
        };
    }

    [Theory]
    [MemberData(nameof(GetTestCases))]
    public void Solve(int l1, int l2, int expected)
    {
        static ListNode ConvertToLinkedList(int num)
        {
            ListNode? head = null, tail = null;
            while (num > 0) {
                var node = new ListNode(num % 10);
                if (tail is null) {
                    head = tail = node;
                }
                else {
                    tail = tail.next = node;
                }
                num /= 10;
            }
            return head ?? new ListNode();
        }

        static int ConvertFromLinkedList(ListNode? head)
        {
            var num = 0;
            var multiple = 1;
            while (head is not null) {
                num += head.val * multiple;
                multiple *= 10;
                head = head.next;
            }
            return num;
        }

        var solution = new Solution();
        var num1 = ConvertToLinkedList(l1);
        var num2 = ConvertToLinkedList(l2);
        Assert.Equal(expected, ConvertFromLinkedList(solution.AddTwoNumbers(num1, num2)));
    }
}

File: rust/src/p0002_add_two_numbers.rs
use crate::linked_list::ListNode;
pub struct Solution;

impl Solution {
    pub fn add_two_numbers(l1: Option<Box<ListNode>>, l2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let default_node = Box::new(ListNode::new(0));
        let mut dummy_head = Some(Box::new(ListNode::new(0)));
        let mut tail_ref = &mut dummy_head;
        let mut l1_ref = &l1;
        let mut l2_ref = &l2;
        let mut carry = 0;
        while l1_ref.is_some() || l2_ref.is_some() || carry > 0 {
            let l1_node = l1_ref.as_ref().unwrap_or(&default_node);
            let l2_node = l2_ref.as_ref().unwrap_or(&default_node);
            let sum = l1_node.val + l2_node.val + carry;
            match *tail_ref {
                None => unreachable!(),
                Some(ref mut tail_node) => {
                    tail_node.next = Some(Box::new(ListNode::new(sum % 10)));
                    tail_ref = &mut tail_node.next;
                }
            }
            l1_ref = &l1_node.next;
            l2_ref = &l2_node.next;
            carry = sum / 10;
        }
        dummy_head.unwrap().next
    }
}

#[cfg(test)]
mod test {
    use test_case::test_case;
    use super::*;

    #[test_case(1, 2, 3)]
    #[test_case(1, 20, 21)]
    #[test_case(1, 999, 1000)]
    fn solve(num1: i32, num2: i32, expected: i32) {
        let l1 = ListNode::from_num_reversed(num1);
        let l2 = ListNode::from_num_reversed(num2);
        assert_eq!(ListNode::to_num_reversed(&Solution::add_two_numbers(l1, l2)), expected);
    }
}